git-worktree/src/lib.rs

```diff
@@ -1 +1,111 @@
 #![forbid(unsafe_code, rust_2018_idioms)]
+//! Git Worktree
+
+use git_hash::oid;
+use git_index::{
+    entry::{Flags, Mode},
+    State,
+};
+use git_object::bstr::ByteSlice;
+use git_object::Data;
+use quick_error::quick_error;
+use std::convert::TryFrom;
+use std::fs;
+use std::fs::create_dir_all;
+use std::path::Path;
+
+#[cfg(unix)]
+use std::os::unix::fs::PermissionsExt;
+
+quick_error! {
+    #[derive(Debug)]
+    pub enum Error {
+        Utf8Error(err: git_object::bstr::Utf8Error) { from() }
+        TimeError(err: std::time::SystemTimeError) { from() }
+        ToU32Error(err: std::num::TryFromIntError) { from() }
+        IoError(err: std::io::Error) { from() }
+    }
+}
+
+/// Copy index to `path`
+pub fn copy_index<P, Find>(state: &mut State, path: P, mut find: Find, opts: Options) -> Result<(), Error>
+where
+    P: AsRef<Path>,
+    Find: for<'a> FnMut(&oid, &'a mut Vec<u8>) -> Option<Data<'a>>,
+{
+    let path = path.as_ref();
+    let mut buf = Vec::new();
+    let mut entry_time = Vec::new(); // Entries whose timestamps have to be updated
+    for (i, entry) in state.entries().iter().enumerate() {
+        if entry.flags.contains(Flags::SKIP_WORKTREE) {
+            continue;
+        }
+        let dest = path.join(entry.path(state).to_path()?);
+        create_dir_all(dest.parent().expect("path is empty"))?;
+        match entry.mode {
+            Mode::FILE | Mode::FILE_EXECUTABLE => {
+                let obj = find(&entry.id, &mut buf).unwrap();
+                std::fs::write(&dest, obj.data)?;
+                if entry.mode == Mode::FILE_EXECUTABLE {
+                    #[cfg(unix)]
+                    fs::set_permissions(&dest, fs::Permissions::from_mode(0o777)).unwrap();
+                }
+                let met = std::fs::metadata(&dest)?;
+                // Set both fields to mtime if ctime is not present and vice-versa
+                // If both fields are not present, will raise an Err
+                let ctime = met.created().or_else(|_| met.modified())?;
```

Comment:
This is well-meant and it would make sense to do that, too. Git unfortunately [doesn't go this far](https://github.com/git/git/blob/main/read-cache.c#L154:L154) and instead just sets these values to zero if the fields do not exist.

That said, if the `metadata()` call succeeds, there will be data for `created()` and `modified()` if the fields exist, and if not, these should just be set to zero. Zero is the value the struct fields would have after a memset.

The reason we do this is to write exactly the same index that git would write, to not alter its behavior if gitoxide and git are used on the same repo.

Reply:
Even if the `metadata` call succeeds, `mtime` or `ctime` may fail. My initial assumption was that they would succeed but `mtime` succeeds on unix and `ctime` fails as unix does not store creation time.

To quote https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap04.html#tag_04_09, 
> Each file has three distinct associated timestamps: the time of last data access, the time of last data modification, and the time the file status last changed

Reply:
It's undisputed that they may fail, it's the fallback which isn't correct as git keeps these values at 0. There is code linked for reference which means that it will just copy whatever is present in a previously 0-initialized struct.